### PR TITLE
Add windows support

### PR DESCRIPTION
**Pull request by tdcosta100**
I have specific needs to run Maplibre GL Native on Windows. So I created a port to Windows platform. It was inspired by Linux platform sources, and works in a similar way, creating some executables and the lib/node-vxx/mbgl.node modules. Therefore, Maplibre GL Native can be used with NodeJS in an Windows environment. I used TileServer-GL to test it, and it worked like a charm, the performance was similar to the Linux version. Build instructions are located at `platform/windows/README.md`. Remember to clone with `--branch windows-support`, because it have a new submodule. Or after changing to `windows-support` branch, you can init the `platform/windows/vendor/vcpkg` submodule manually. Feedback is welcome.

### CODE REVIEW

**Comment by wipfli** [comment]
Oh beautiful, thanks @tdcosta100 for implementing windows support. 

If we were to merge this, would you be interested in being the lead-maintainer for the Windows platform binding?

The usual work consists of handling dependency updates, reviewing pull requests, and helping people getting started with the platform binding...

File: CMakeLists.txt
Comment (ntadej) [inline CMakeLists.txt:17]: This looks a bit intrusive. Qt for example should not use this toolchain. Is it too late if you load it in the platform-specific cmake file?

**Comment by tdcosta100** [comment]
> Oh beautiful, thanks @tdcosta100 for implementing windows support.
> 
> If we were to merge this, would you be interested in being the lead-maintainer for the Windows platform binding?
> 
> The usual work consists of handling dependency updates, reviewing pull requests, and helping people getting started with the platform binding...

I can try. I'm just afraid sometimes I will be with lots of demand from my job and not respond so quickly as I would like. Outside these situations, I think I can do the job. Can we make a trial period to see if I am a good fit for it?

File: benchmark/CMakeLists.txt
Comment (ntadej) [inline benchmark/CMakeLists.txt:57]: Please fix the warnings in non-vendor files.

File: platform/windows/windows.cmake
Comment (ntadej) [inline platform/windows/windows.cmake:2]: I would not split the file per compiler.

File: platform/windows/include/unistd.h
Comment (ntadej) [inline platform/windows/include/unistd.h:1]: I would make clear which files are just platform-specific implementations for MapLibre and which are replacements for system headers.

Also I'm not sure if I like we're hacking system-like headers. How many times do we actually need functions from this file?

**Review by ntadej — CHANGES_REQUESTED** [review]
Thank you very much for providing this!

I provided some basic comments, I may look through the new files again later. To get this merged faster I would recommend to split the MR into smaller ones (e.g. fixing errors).

It would also be good to add a CI workflow to test your addition actually works.

**Comment by tdcosta100** [comment]
> Thank you very much for providing this!
> 
> I provided some basic comments, I may look through the new files again later. To get this merged faster I would recommend to split the MR into smaller ones (e.g. fixing errors).
> 
> It would also be good to add a CI workflow to test your addition actually works.

No worries about the changes. I can change everything, even open a new PR if it's better. I would propose this project to be evaluated before doing a PR, but I just didn't know where to do that hahaha. So if it's not mature enough, I will make the necessary changes before the merge.

File: CMakeLists.txt
Comment (tdcosta100) [reply CMakeLists.txt:17]: I tried to make every if(WIN32) contained in the windows.cmake file, but this toolchain file needs to be included before the project command, otherwise it will be completely ignored. The idea here is to inject vcpkg if we are going to build in Windows platform. I'm heavily using vcpkg to simplify the dependencies.

File: benchmark/CMakeLists.txt
Comment (tdcosta100) [reply benchmark/CMakeLists.txt:57]: Just to better understand the ability to change files: it's ok to make source code modifications instead supressing warnings? Some places have changes like this piece of code, but I can have the #if _MSVC inside the source code approach as well.

Comment (ntadej) [reply benchmark/CMakeLists.txt:57]: I looked up `C4244` which is "floating point truncation". This one can be fixed for all platforms. As such fixes caused regressions in the past, it may be useful to open them in a separate PR.

File: platform/windows/include/unistd.h
Comment (tdcosta100) [reply platform/windows/include/unistd.h:1]: I can check which functions from this code are used. All I can say by now is some files are including unistd.h. We can trace back these includes and make some deeper changes than this hack.

File: platform/windows/windows.cmake
Comment (tdcosta100) [reply platform/windows/windows.cmake:2]: I was hoping to have MingGW support in the future. If so, it would make sense? Or it's better even with MinGW support to have just one file?

Comment (ntadej) [reply platform/windows/windows.cmake:2]: For Qt the changes between MSVC and MinGW are quite small. Also it will probably cause less duplication of code.

File: benchmark/CMakeLists.txt
Comment (tdcosta100) [reply benchmark/CMakeLists.txt:57]: That's nice, I was hoping this change would be welcome. There are are some of these in several files. Using static_cast would make it better for all platforms. I will provide a separate PR to address this.

Comment (ntadej) [reply benchmark/CMakeLists.txt:57]: To clarify, I fixed most of those when I did initial Qt port to Windows, but not all files are used there. We only ignore warnings for external packages where we have no direct control.

File: platform/windows/windows.cmake
Comment (tdcosta100) [reply platform/windows/windows.cmake:2]: Great. So I will migrate windows.msvc.cmake to windows.cmake, and when time comes, make the required changes to have it compatible with MinGW.

File: benchmark/CMakeLists.txt
Comment (tdcosta100) [reply benchmark/CMakeLists.txt:57]: Thanks for the clarification. Now I have a better understanding where I should touch the code, and where I souldn't. My primary concern was to be the least intrusive as I could. But if I have the permission to change some pieces of code (nothing too extensive, just some fixes like this one), I'd rather to.

**Comment by wipfli** [comment]
> can I setup a workflow in a non-main branch

Yes, like this for example

```yaml
on:
  workflow_dispatch:
  pull_request:
    branches:
      - main
      - some_other_branch
```

**Comment by acalcutt** [comment]
If you need any help with getting this into the node-ci/build, let me know. It would be nice to add node windows support.

If you modify the node-ci workflow in the PR it is already set to run on pull request, so I don't think you would need to modify the branches it runs on. (at least that was how I was testing it before it got merged)

**Comment by tdcosta100** [comment]
> > can I setup a workflow in a non-main branch
> 
> Yes, like this for example
> 
> ```yaml
> on:
>   workflow_dispatch:
>   pull_request:
>     branches:
>       - main
>       - some_other_branch
> ```

Nice, thanks! I will do that then.

**Comment by tdcosta100** [comment]
Once all pending changes to prepare for this PR are finally done (that is, changes that are not directly related with Windows, like Khronos definitions, and implicit casts), I'm resuming this PR. I converted it to draft in order to test actions for `NodeJS` on Windows platform. I didn't include actions yet, but once all tests passed, I will focus that.

File: platform/windows/FindOSMesa.cmake
Comment (acalcutt) [inline platform/windows/FindOSMesa.cmake:15]: will the eventually need a arm64 if we want to support that?

**Comment by acalcutt** [comment]
I spent some time working with @tdcosta100 to try and get this integrated into the 'node-ci' and 'node-release' workflows

We were able to get it all working, but I wanted to note a few issues we had along the way.

1.) When checking out the the repository on the github windows runner, I ran into path limit issues with some vendor test files. To get around this, I took a solution from the 'qt-ci', and did a shallow checkout and then pull the submodules after. This gets around the issue, but you can still see some errors relating to the paths that don't affect the build.

2.) To get ccache working in windows, I again pulled from the qt-ci, and use that to simplify our ccache use here (and make it work with windows).

3.) We initially had some issue with build type not being release, which made it this require c++ development dlls. we fixed that issue, but it should still be noted this requires c++ redistributable as a pre-req 

Other than that, I was able to ci/release workflow that works. 

I made a test release at
`NPM: "@acalcutt/maplibre-gl-native": "6.1.0-pre.4"`

and a tileserver-gl that uses it at 
https://github.com/acalcutt/tileserver-gl/tree/windows_test

With that I was able to confirm this works!
image
image

Great work @tdcosta100

**Comment by tdcosta100** [comment]
Regarding the too long path issues, I confirm it's resolved, at least when cloning submodules (recursively!). I don't know how tests will behave because I didn't let a test job finish yet, but I hope in the next few hours we will confirm if it works correctly.

**Comment by tdcosta100** [comment]
Finally got Windows tests working. Thank you @acalcutt again, for helping with the workflow.

File: platform/windows/FindOSMesa.cmake
Comment (tdcosta100) [reply platform/windows/FindOSMesa.cmake:15]: I don't see why not, but I really don't have tools for developing that. But theorically yes, it's possible.

File: CMakeLists.txt
Comment (tdcosta100) [reply CMakeLists.txt:17]: Do you think it's better to have the custom toolchain file to be included in cmake command? Something like:

```
cmake . -B build -DCMAKE_TOOLCHAIN_FILE=platform/windows/custom-toolchain.cmake
```

File: render-test/filesystem.hpp
Comment (louwers) [inline render-test/filesystem.hpp:3]: Can this be set with CMake instead?

File: platform/default/src/mbgl/storage/http_file_source.cpp
Comment (louwers) [inline platform/default/src/mbgl/storage/http_file_source.cpp:284]:
```suggestion
    // Windows has issues with TLSv1.3, so we limit to TLSv1.2. Should be resolved in a later cURL release
```

File: expression-test/filesystem.hpp
Comment (louwers) [inline expression-test/filesystem.hpp:4]: Can this `#define` be added with CMake?

**Comment by louwers** [comment]
Added a new label for windows. 😉

**Comment by louwers** [comment]
Could you update the root `README.md`:

```
- [⭐️ Android](platform/android/README.md) 
- [⭐️ iOS](platform/ios/platform/ios/README.md)
- [GLFW](platform/glfw)
- [Linux](platform/linux/README.md)
- [Node.js](platform/node/README.md)
- [Qt](platform/qt/README.md)
- [Windows](platform/windows/README.md)
- [macOS](platform/ios/platform/macos/README.md)
```

**Comment by etnav** [comment]
This is awesome. Thank you @tdcosta100 (and the others of course) for your effort, I can relate to your struggles. Looking forward to seeing this merged and migrating to this (built this on windows about 2 years ago), maybe even contributing to this specific part in the future😄

File: .github/workflows/node-ci.yml
Comment (acalcutt) [inline .github/workflows/node-ci.yml:170]: can this be moved back to bash with the other syntax. the ccache plugin we are using notes

```
Notes on Windows support
Note that using Ccache on Windows is still experimental and https://github.com/ccache/ccache/issues/1023. Use Sccache for stable Windows support.
```
https://github.com/hendrikmuhs/ccache-action#notes-on-windows-support

**Review by ovivoda — APPROVED** [review]

File: CMakeLists.txt
Comment (ntadej) [reply CMakeLists.txt:17]: Sorry to unresolve this now, but this is not OK for Qt builds, even if it may work by accident.

File: expression-test/filesystem.hpp
Comment (ntadej) [reply expression-test/filesystem.hpp:4]: +1

File: platform/node/cmake/module.windows.cmake
Comment (ntadej) [inline platform/node/cmake/module.windows.cmake:1]: This comment does not fit Windows 😊

File: platform/node/cmake/module.windows.cmake
Comment (ntadej) [inline platform/node/cmake/module.windows.cmake:5]: Why do you need so many Node specifics for Windows but not for other platforms?

**Review by ntadej — CHANGES_REQUESTED** [review]

File: platform/node/cmake/module.windows.cmake
Comment (acalcutt) [reply platform/node/cmake/module.windows.cmake:5]: I asked about this file, and the linux/macos equivalent is in the `mapbox/cmake-node-module` package
https://github.com/mapbox/cmake-node-module/blob/master/module.cmake

We talked a bit about if it may make sense to move the one for other platforms out of the cmake-node-module package so we have some control over it.

Comment (ntadej) [reply platform/node/cmake/module.windows.cmake:5]: OK, I'm fine with including this for now then.

Comment (acalcutt) [reply platform/node/cmake/module.windows.cmake:5]: Just a side fact. The linux version of the file being in a node package actually gave me some trouble in my cmake-js PR, because the path of node_modules changes if maplibre-native is used as a submodule

https://github.com/maplibre/maplibre-gl-native/pull/535/files#diff-6d48556e62e5d1ff41c30848ca089f0bb2fe15ddf3b58bcc1f33e902b7022eddR1-R6

Comment (acalcutt) [reply platform/node/cmake/module.windows.cmake:5]: @tdcosta100 just wondering, would this new file also be compatible with the linux/macos builds. like maybe replace module.cmake with this new file all together?

Looking at the difference it seems like mostly a few path changes and sections for WIN32 added.

**Comment by tdcosta100** [comment]
> Could you update the root `README.md`:
> 
> ```
> - [⭐️ Android](platform/android/README.md) 
> - [⭐️ iOS](platform/ios/platform/ios/README.md)
> - [GLFW](platform/glfw)
> - [Linux](platform/linux/README.md)
> - [Node.js](platform/node/README.md)
> - [Qt](platform/qt/README.md)
> - [Windows](platform/windows/README.md)
> - [macOS](platform/ios/platform/macos/README.md)
> ```

Done!

File: README.md
Comment (acalcutt) [inline README.md:30]: I feel like macOS\Linux\Windows should be together.

File: platform/node/cmake/module.cmake
Comment (acalcutt) [inline platform/node/cmake/module.cmake:138]: the indenting in this doesn't seem to match the lines above/below

File: platform/node/cmake/module.cmake
Comment (acalcutt) [inline platform/node/cmake/module.cmake:169]: should this only apply to WIN32?

File: CMakeLists.txt
Comment (acalcutt) [reply CMakeLists.txt:17]: would adding "AND NOT MBGL_WITH_QT" resolve this or would you want something else?

Comment (ntadej) [reply CMakeLists.txt:17]: Toolchain files should generally not be included by default and can be just provided when you call CMake (this is for example usually done for Android and iOS). Can you rather adapt your Node.js building scripts in such a way?

Comment (acalcutt) [reply CMakeLists.txt:17]: Is that like @tdcosta100  suggested a few posts above with specifying DCMAKE_TOOLCHAIN_FILE? Can you point to an example from the Android/iOS that we could see what you mean? I looked a bit through those worksflows but I didn't really see anything that looked like what you describe.

Comment (acalcutt) [reply CMakeLists.txt:17]: really just looking at what can be done to finish this PR off. Maybe another switch that's off by default that we can turn on witht he cmake command.

Comment (acalcutt) [reply CMakeLists.txt]: @ntadej , this has been changed in the current version to not apply to QT. So I'm still looking if this resolve this for you or what you would like instead. This has already been approved and tested, so please this is the one thing in the way of merge I would say.

Comment (ntadej) [reply CMakeLists.txt]: Indeed we can clean it up later.

**Review by ntadej — APPROVED** [review]

**Pull request merged**
